nuxeo-features/nuxeo-thumbnail/src/main/java/org/nuxeo/ecm/platform/thumbnail/action/RecomputeThumbnailsAction.java

```diff
@@ -0,0 +1,88 @@
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Contributors:
+ *     Nour Al Kotob
+ */
+
+package org.nuxeo.ecm.platform.thumbnail.action;
+
+import static org.nuxeo.ecm.core.bulk.BulkServiceImpl.STATUS_STREAM;
+import static org.nuxeo.ecm.platform.thumbnail.ThumbnailConstants.THUMBNAIL_PROPERTY_NAME;
+import static org.nuxeo.lib.stream.computation.AbstractComputation.INPUT_1;
+import static org.nuxeo.lib.stream.computation.AbstractComputation.OUTPUT_1;
+
+import java.io.Serializable;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+
+import org.nuxeo.ecm.core.api.Blob;
+import org.nuxeo.ecm.core.api.CoreSession;
+import org.nuxeo.ecm.core.api.DocumentModel;
+import org.nuxeo.ecm.core.api.DocumentModelList;
+import org.nuxeo.ecm.core.api.thumbnail.ThumbnailAdapter;
+import org.nuxeo.ecm.core.bulk.action.computation.AbstractBulkComputation;
+import org.nuxeo.ecm.core.bulk.message.BulkCommand;
+import org.nuxeo.lib.stream.computation.Topology;
+import org.nuxeo.runtime.stream.StreamProcessorTopology;
+
+/**
+ * Bulk Action processor that generates thumbnails from documents.
+ *
+ * @since 11.1
+ */
+public class RecomputeThumbnailsAction implements StreamProcessorTopology {
+
+    public static final String ACTION_NAME = "recomputeThumbnails";
+
+    public static final String PARAM_XPATH = "xpath";
+
+    @Override
+    public Topology getTopology(Map<String, String> options) {
+        return Topology.builder()
+                       .addComputation(RecomputeThumbnailsComputation::new, //
+                               Arrays.asList(INPUT_1 + ":" + ACTION_NAME, OUTPUT_1 + ":" + STATUS_STREAM))
+                       .build();
+    }
+
+    public static class RecomputeThumbnailsComputation extends AbstractBulkComputation {
+
+        protected String xpath;
+
+        public RecomputeThumbnailsComputation() {
+            super(ACTION_NAME);
+        }
+
+        @Override
+        public void startBucket(String bucketKey) {
+            BulkCommand command = getCurrentCommand();
+            xpath = command.getParam(PARAM_XPATH);
+        }
+
+        @Override
+        protected void compute(CoreSession session, List<String> ids, Map<String, Serializable> properties) {
+
+            DocumentModelList docs = loadDocuments(session, ids);
+
+            for (DocumentModel doc : docs) {
+                ThumbnailAdapter thumbnail = doc.getAdapter(ThumbnailAdapter.class);
+                Blob thumbnailBlob = thumbnail.computeThumbnail(session);
+                doc.setPropertyValue(THUMBNAIL_PROPERTY_NAME, (Serializable) thumbnailBlob);
+                thumbnail.save(session);
```

Comment:
I think you should better take the whole code from the `UpdateThumbnailListener#processDoc` method.

Reply:
Took it, took the submethod, took the related constant and made the methods static to work